admiral/envs/agent_based_simulation.py

```diff
@@ -1,6 +1,129 @@
+        assert value is None or type(value) is int, "Seed must be an integer."
+        self._seed = value
+
+    @property
+    def configured(self):
+        """
+        Determine if the agent has been successfully configured.
+        """
+        return self.id is not None
+    
+    def finalize(self, **kwargs):
+        pass
+
+    def __eq__(self, other):
+        return self.__dict__ == other.__dict__ if isinstance(other, self.__class__) else False
+
+class ActingAgent(Agent):
+    """
+    ActingAgents are Agents that are expected to produce actions and therefore
+    should have an action space in order to be successfully configured.
+    """
+    def __init__(self, action_space=None, **kwargs):
+        super().__init__(**kwargs)
+        self.action_space = action_space
+    
+    @property
+    def action_space(self):
+        return self._action_space
+    
+    @action_space.setter
+    def action_space(self, value):
+        assert value is None or gu.check_space(value), \
+            "The action space must be None, a gym Space, or a dict of gym Spaces."
+        self._action_space = {} if value is None else value
+    
+    @property
+    def configured(self):
+        return super().configured and gu.check_space(self.action_space, strict=True)
+            
+    def finalize(self, **kwargs):
+        """
+        Wrap all the action spaces with a Dict if applicable and seed it if the agent was
+        created with a seed.
+        """
+        super().finalize(**kwargs)
+        if type(self.action_space) is dict:
+            self.action_space = gu.make_dict(self.action_space)
+        self.action_space.seed(self.seed)
+
+class ObservingAgent(Agent):
+    """
+    ObservingAgents are Agents that are expected to receive observations and therefore
+    should have an observation space in order to be successfully configured.
+    """
+    def __init__(self, observation_space=None, **kwargs):
+        super().__init__(**kwargs)
+        self.observation_space = observation_space
+
+    @property
+    def observation_space(self):
+        return self._observation_space
+    
+    @observation_space.setter
+    def observation_space(self, value):
+        assert value is None or gu.check_space(value), \
+            "The observation space must be None, a gym Space, or a dict of gym Spaces."
+        self._observation_space = {} if value is None else value
+    
+    @property
+    def configured(self):
+        return super().configured and gu.check_space(self.observation_space, strict=True)
+
+    def finalize(self, **kwargs):
+        """
+        Wrap all the observation spaces with a Dict and seed it if the agent was
+        created with a seed.
+        """
+        super().finalize(**kwargs)
+        if type(self.observation_space) is dict:
+            self.observation_space = gu.make_dict(self.observation_space)
+        self.observation_space.seed(self.seed)
+
+class PureAgent(ObservingAgent, ActingAgent):
```

Comment:
PureAgent is a dumb name. Consider the following possibilities:
1. PrincipleAgent
2. CoreAgent
3. BasisAgent
4. BaseAgent
5. SimpleAgent

Also consider using these names for the super Agent class and just naming this one Agent.